### ISSUE TITLE
Make the output channel generic

### ISSUE THREAD

**Issue by vincent-herlemont** [vincent-herlemont/native_db#249]
Currently, event listening via the `watch` command does not support generic channel management. The user is forced to choose between the `tokio` feature or the standard Rust channel (see: [src/watch/mod.rs#L38-L46](https://github.com/vincent-herlemont/native_db/blob/bcf1c6bd0c4e669b2b4385005aae3e4efaffcd55/src/watch/mod.rs#L38-L46)). The purpose of this issue is to investigate whether it is possible to elegantly support any type of channel.

# TODO
- [ ] Investigate

**Comment by lu-zero** [comment]
The less than elegant solution is to rely on [flume](https://crates.io/crates/flume) or another crate supporting both sync and async at the same time. If you think it is acceptable I can provide a PR

**Comment by vincent-herlemont** [comment]
@lu-zero this could save a feature. However, the goal here is to find a way to be compatible with all types of channels to avoid using multiple types of channels in a crate.

**Comment by lu-zero** [comment]
There isn't a common trait across all the channel implementation so there isn't a way even to support alternative implementations of sync channels like flume vs crossbeam vs std or async channels such as flume vs tokio vs async-channel ^^;

**Comment by vincent-herlemont** [comment]
Yes, you are right, there won't be any generics in terms of Rust declarations. However, we could instead find a way to make it "generic" by using a [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html) (that takes the event as an argument), which will be defined by the users, as a parameter and is called every time an event is sent.

**Comment by lu-zero** [comment]
Storing `Fn` requires a generic and tends to be annoying in my experience though...

**Comment by vincent-herlemont** [comment]
> Storing Fn requires a generic

@lu-zero How does this imply a generic declaration? Here's what I imagined, what's your point of view?

```rust
// /!\ Pseudo code
let (sender, receiver) = mpsc::channel();
db.watch(move |event| {
    // Send the event message into the channel
   sender.send(event).unwrap();
});
```

As you can see, you can put whatever you want in the closure. This does not imply a generic.

**Comment by vincent-herlemont** [comment]
@lu-zero Can you provide a detailed example of a situation where this would pose a problem?

**Comment by lu-zero** [comment]
It is easier to write down the signature of your pseudocode I guess :)

```rust
// `clo` is generic over anything implementing `FnMut`
fn watch(&mut self, clo: impl FnMut(EventTy)) {

// and if we want to save it someplace we have to box it
self.closure = Box::new(clo);

...
}  
```

This should be a way to make it work for sync, for async it gets more annoying since we do not have [yet](https://blog.rust-lang.org/inside-rust/2024/08/09/async-closures-call-for-testing.html) async closures.

**Comment by vincent-herlemont** [comment]
@lu-zero Yes, that's exactly the idea! And indeed, for async, we'll have to wait a bit ;)

If you need support for [flume](https://crates.io/crates/flume) or something else in the meantime, you can add it to the two existing channel supports, and I will merge it.

**Comment by lu-zero** [comment]
I do not need it myself, but since it is just few minutes to implement it using flume, it could solve the use-case of keeping the same API across features.
